### ISSUE TITLE
Add a .zip version for the Win32/64 nightly builds

### ISSUE THREAD

**Issue by justinclift** [sqlitebrowser/sqlitebrowser#1630]
Following up the mention of this in https://github.com/sqlitebrowser/sqlitebrowser/issues/1613#issuecomment-441799983.  It should be fairly easy for us to get done now that we have the new .msi builds happening + the build scripts are fairly simple. :smile:

The wiki page @karim created with instructions is here:

&nbsp; &nbsp; https://github.com/sqlitebrowser/sqlitebrowser/wiki/Extracting-installer-contents

In theory, it will mean adding the msi extraction command to our build script, then something to create the .zip file.  Something along the lines of:

```
msiexec /a "C:\builds\DB Browser for SQLite-sqlcipher-%RUN_DATE%-win64.msi" /q TARGETDIR=C:\builds\extracted
"C:\Program Files\7-Zip\7z.exe" -a -r -y C:\builds\DB Browser for SQLite-sqlcipher-%RUN_DATE%-win64..zip C:\builds\extracted\
```

Note that's untested so far.  Shouldn't be hard to try it out and make it work decently though. :smile:

**Comment by karim** [comment]
You beat me to it. :smiley: 

I was testing this yesterday and I got it working with 64-bit build. I was waiting to test it with 32-build then make a PR with it.

Basically it is something like this. After building the .msi inside `C:\builds` ...

```
CD %DEST_PATH%
msiexec /a "DB Browser for SQLite-%RUN_DATE%-win64.msi" /q TARGETDIR=%CD%\zip
MOVE %CD%\zip\System64\* "%CD%\zip\DB Browser for SQLite"
"C:\Program Files\7-Zip\7z.exe" a "DB Browser for SQLite-%RUN_DATE%-win64.zip" "%CD%\zip\DB Browser for SQLite"
RMDIR /S /Q %CD%\zip
```

**Comment by karim** [comment]
Yes, correct.

On the build machine, do you run the same scripts inside `installer/windows/nightly_build_scripts` folder? Or they are just copies? I'm asking because I was going to edit these files and wasn't sure if the build system will pick them up. Also, where are the other build scripts for other systems, like Mac?

**Comment by justinclift** [comment]
The ones in the repo are just copies.  They're not run live out of a git checkout, so changes won't be picked up.  Feel free to edit them though, and I'll copy them across to the place they're really run from. :smile:

For the mac ones, I copied the build script into the git repo earlier today. :grinning:

**Comment by karim** [comment]
I see, thanks.

> For the mac ones, I copied the build script into the git repo earlier today. :grinning:

Should we also move `snap` folder to `installer/linux/snap` ?

**Comment by justinclift** [comment]
Sounds ok to me, but there may be a need for it to be in the root of the repo.  Not sure, lets check with @deepsidhu1313 before moving it, just in case. :smile:

**Comment by karim** [comment]
I updated the build scripts now. Can you check them please, cause I'm not sure about the last part, where the files get uploaded to the nightly server. :smiley:

**Comment by deepsidhu1313** [comment]
Thank you @justinclift for mentioning me here. Yes snap dir there is on purpose. I can do research into that, if build tools allow us to move snap directory. But most likely, they need it there.

**Comment by justinclift** [comment]
@deepsidhu1313 No worries.  We can just leave it in the root directory.  It's just @karim and I thinking about making things neater, since we're already working on other build pieces yesterday and today anyway.  No desire to break things though. :smile:

**Comment by justinclift** [comment]
@karim The updated script pieces look ok to me.  I'll copy them to the place they're run from, so we can watch them ~explode then delete the server~ work perfectly fine tomorrow. :stuck_out_tongue_winking_eye:

**Comment by karim** [comment]
@justinclift :smile: 

@deepsidhu1313 Thanks for clarifying that. It was just for keeping things tidy.

**Comment by karim** [comment]
@justinclift Maybe you can update https://nightlies.sqlitebrowser.org/latest/%2A%20README.txt to include info about the new .zip files when they are uploaded?

**Comment by justinclift** [comment]
Good point.  I'd forgotten about that. :smile:

**Comment by karim** [comment]
@justinclift While we are at it, do you think we should merge both .msi, the one with normal SQLite and the one with SQLCipher?

The only difference between both versions are 4 files. Two .dlls (`sqlite3.dll` and `sqlcipher.dll`) that can be installed together (or only one of them), and two .exe that we can choose one to install during installation.

We install the SQLite version by default (to get the latest version), while keeping a checkbox to install/add encryption, and in that case we install the SQLCipher .exe instead. The only downside is the installer size increase.

Do you think it is worth the effort or it will be confusing to others?

**Comment by justinclift** [comment]
> @justinclift While we are at it, do you think we should merge both .msi, the one with normal SQLite and the one with SQLCipher?

Interesting idea.  I'd prefer for us to provide the encryption-enabled one by default, but if you can add some kind of option in the install dialog which lets people choose the no-encryption version, then sure.  That would probably let us merge things together in a fairly optimal way. :smile:

**Comment by karim** [comment]
I said SQLite by default since we can always include the latest version. I don't know the stats of the downloads, but I think most users don't use encryption, but I could be wrong. I think you are right, encryption-enabled by default sounds like the best practice.

Another approach is installing both .exe (since they are included in the installer anyway) after renaming one of them. This way, you can have, and use, both together at the same time. This will allows us to create one .zip file also.

We just need two different names for each .exe, maybe..
- `DB Browser for SQLite.exe`
- `DB Browser for SQLCipher.exe`

And on the Desktop, we include two shortcuts..
- `DB Browser for SQLite` pointing to `DB Browser for SQLite.exe`
- `DB Browser for SQLite (with encryption)` pointing to `DB Browser for SQLCipher.exe`

**Comment by SilvioGrosso** [comment]
Hello @karim @justinclift 

While you are overhauling the Windows installers (.msi, .zip etc) I am wondering whether it would be possible to include also the "Math extension" [1] in order to make it available in the upcoming stable release.
This would be useful in order to have some time to test this extension on Windows (10) to make sure it works properly :-)

[1] https://github.com/sqlitebrowser/sqlitebrowser/issues/1224

**Comment by justinclift** [comment]
Oops, good reminder.  I keep on forgetting about that.

In theory (!), if we can figure out how to compile it then making sure it's included should be fairly straight forward.  Lets try and get it done for beta2. :smile:

**Comment by justinclift** [comment]
@karim While we at it, we also need to figure out how to have the .msi installer notice previous (non-MSI) installed versions of DB4S, so it can uninstall those.  Would you have time to look into it?

**Comment by karim** [comment]
> I am wondering whether it would be possible to include also the "Math extension"

@SilvioGrosso There are a lot of useful extensions, and I don't know if we should add them all to the installer or create a different package for each one that can be installed on top of DB4S. We will have to test both and see.

> @karim While we at it, we also need to figure out how to have the .msi installer notice previous (non-MSI) installed versions of DB4S, so it can uninstall those. Would you have time to look into it?

@justinclift Ahh yes. I completely forgot about that, sorry. I will look into it tomorrow when I'm free.

**Comment by justinclift** [comment]
@karim No worries. :smile:

Found the main tracking issue for the math extension too, it's #1224.

**Comment by justinclift** [comment]
Just finished running the updated Win nightly build batch files.  The new .zip creation bits seem to work fine, and the uploads look good too.  Well done @karim! :smile:

**Comment by SilvioGrosso** [comment]
Hello everyone (Karim, Justin),
As regards the math extension, I agree it might be installed later on by the user himself.However, at least on Windows (10), I suppose it is not that easy for laymen to do so. Needless to say, I would be happy to be proved wrong :-)

Compared to others Sqlite extensions, IMHO, this one add some very basic functions, both for handling texts and numbers and this is the main reason I have been pestering you so much lately for its inclusion :-)
 

    Il Martedì 27 Novembre 2018 23:09, Justin Clift <notifications@github.com> ha scritto:
 

 @karim While we at it, we also need to figure out how to have the .msi installer notice previous (non-MSI) installed versions of DB4S, so it can uninstall those. Would you have time to look into it?—

**Comment by chrisjlocke** [comment]
> However, at least on Windows (10), I suppose it is not that easy for laymen to do so. 

The extension has been kindly compiled by @struck89, so all you have to do is click 'Tools' and 'Load Extension'.  Its quite easy.  

![image](https://user-images.githubusercontent.com/7872218/49147281-29b1d200-f2fc-11e8-98f3-104dcbcaaa1e.png)

**Comment by SilvioGrosso** [comment]
Hello @chrisjlocke ,

First off, thanks a bunch *indeed* for your help: It is highly appreciated :-)

Just tried to downlaod the .ddl through the link you suggested (https://github.com/struck89).
Sorry for being dumb but I am unable to download the compiled math extension in order to try to import it into Db Browser. 
Actually, I was aware of the option to get them directly from the Db Browser GUIs but I thought  this extension was not available on Windows out of the box.
Glancing at the repositories it looks like none of them (Python, MathLab and the like) is what I am looking for :-)
I attach a screenshot as regards the stuck89/ repositories you pointed me out where I got lost in trying to pick the correct one:

![github png](https://user-images.githubusercontent.com/33007383/49179811-6826ab80-f353-11e8-9989-fd7e841ff0c6.jpg)

BTW, I took a gander at the Sqlite official web page where the math extension is originally listed (https://www.sqlite.org/contrib), however,  it looks like I can not pick this one either because it needs to be previously compiled (but, again, I may be totally wrong....)

**Comment by karim** [comment]
@SilvioGrosso The compiled extension download link was in the other issue. Download the .zip file in this comment https://github.com/sqlitebrowser/sqlitebrowser/issues/1224#issuecomment-351681244, extract it, and load it. I think this is what @chrisjlocke meant.

**Comment by SilvioGrosso** [comment]
Hello @karim 

Yep. It does work now!
Many *THANKS* indeed

Here is the screenie for the stdev (standard deviation) which is now possible to yield thanks to this extension (Windows 10 - 64 bit ):

![stdev png](https://user-images.githubusercontent.com/33007383/49182361-5e547680-f35a-11e8-8865-e0db062b5a85.jpg)

**Comment by developernotes** [comment]
Hi @justinclift 

We are happy to provide approval for this project, with regard to the name usage of SQLCipher. We are just waiting on feedback on how to properly grant your project permission to utilize the name SQLCipher.

Out of curiosity, is there a reason for shipping two different binaries, one built with SQLCipher, the other with standard SQLite?

**Comment by justinclift** [comment]
@developernotes That's pretty awesome, thank you! :smile:

With our builds, we generally provide the SQLCipher enabled one by default just so everyone has the ability to encrypt/decrypt stuff "at hand".  And most people don't seem to really need the features in the latest & greatest releases of SQLite.  It's ok for them to use the SQLCipher provided SQLite features.

That being said, there **are** a bunch of our users who do use the latest and greatest features of SQLite.  For them, having very recent / newest SQLite is more important than having encryption.

So, this is us wanting to cater to both groups. :smile:

**Comment by justinclift** [comment]
@karim I'll run the new win32/64 .bat files you created in a few minutes. :smile:

**Comment by karim** [comment]
@justinclift Nice. :smiley: 

If you found any problems, I think it is better to create a new issue. This issue should be closed when we are done here. :smiley:

**Comment by karim** [comment]
> Oh, we still need to update the README.txt file on the nightlies server. Lets get that done when we have this new dual-build thing in place. :smile:

Don't forget to add some info about the .zip package too. :smiley:

**Comment by justinclift** [comment]
The new Win32 dual build (with both SQLite and SQLCipher included) is on the nightly build website now too:

&nbsp; &nbsp; https://nightlies.sqlitebrowser.org/latest/DB.Browser.for.SQLite-win32.msi

Seems ok to me in very initial testing.  Haven't tried to encrypt/decrypted anything using the Win32/64 SQLCipher variants though, so having someone clueful verify they're ok would be good.

Um @developernotes, any interest in taking a look at these two nightly builds (our first with both SQLCipher & SQLite in one .msi)?

* https://nightlies.sqlitebrowser.org/latest/DB.Browser.for.SQLite-win32.msi
* https://nightlies.sqlitebrowser.org/latest/DB.Browser.for.SQLite-win64.msi

We're definitely also open to suggestions for adjusting how bits are named, where shortcuts go, and that kind of thing.  This is our first tentative build like this, just to examine feasibility. :smile:

**Comment by developernotes** [comment]
Hi @justinclift 

I'd definitely like to take a look at them further. At the moment we are finishing up the push for the next SQLCipher release so I won't be able to get to them until we get that out the door and the dust settles!

**Comment by justinclift** [comment]
No worries at all @developernotes.  We'll likely be using that updated SQLCipher in our main release too. :grinning:

**Comment by justinclift** [comment]
Excellent. Just saw that SQLCipher 4.0.0 has been released:

&nbsp; &nbsp; https://www.zetetic.net/blog/2018/11/30/sqlcipher-400-release/

I'll give that a go tomorrow, and if successful push out a new beta with it. :smile:

**Comment by justinclift** [comment]
Actually, that might not be very smart.  Looking at the [CHANGELOG](https://github.com/sqlcipher/sqlcipher/blob/5e991ea68690b8842c47d27bfc7b31e5641b11c4/CHANGELOG.md#400---2018-11-30), it would be better for us to check if any of the changes (eg deprecated PRAGMA's, changed default for block size, etc) mean we should adjust our code to suit first.

The addition of support for OpenSSL's 1.1 series sounds useful too. :smile:

**Comment by karim** [comment]
From your first link.

> Many of these changes provide a much higher level of security than previous versions of SQLCipher. However, the new page size of 4096, 256,000 KDF iterations, use of PBKDF2-HMAC-SHA512 and HMAC-SHA512 all modify important database settings. Thus, **SQLCipher 4 will not open older databases by default.**

You are right. We cannot use this new version without changing our code as well.

**Comment by justinclift** [comment]
Ouch, that sounds pretty blocker-ish.  We'll need to figure out some way to get the SQLCipher version we include to also open existing databases.  That's really not optional. :wink:

**Comment by justinclift** [comment]
As a first initial guess for updating our OpenSSL version to 1.1.x, we'll need to change the library name given on our Windows batch files command line:

* https://github.com/sqlitebrowser/sqlitebrowser/blob/d0417eeac9a06cc2a7bdfb75f3bd414911563705/installer/windows/nightly_build_scripts/win32build_new.bat#L78
* https://github.com/sqlitebrowser/sqlitebrowser/blob/d0417eeac9a06cc2a7bdfb75f3bd414911563705/installer/windows/nightly_build_scripts/win64build_new.bat#L78
  * `libeay32.lib` → `libcrypto.lib`

We'll also need to update the names of the two libraries we copy, in CMakeLists.txt:

* https://github.com/sqlitebrowser/sqlitebrowser/blob/d0417eeac9a06cc2a7bdfb75f3bd414911563705/CMakeLists.txt#L472-L473
  * `libeay32.dll` → `libcrypto-1_1.dll` (32-bit) & `libcrypto-1_1-x64.dll` (64-bit)
  * `ssleay32.dll` → `libssl-1_1.dll` (32-bit) & `libssl-1_1-x64.dll` (64-bit)

Kind of thinking that for our 3.11.0 release we might want to use the older SQLCipher v3.4.2 + OpenSSL 1.0.x.  That would allow time for us to figure out what needs doing for SQLCipher v4.

The only things that spring to mind that we're waiting on for the 3.11.0 release are:

* Getting the MSI installers to detect and correctly remove installers done using the previous .exe releases
* Any outstanding translations

**Comment by justinclift** [comment]
Started experimenting with changes for OpenSSL 1.1 support in a new set of `openssl11_vx` branches:

&nbsp; &nbsp; https://github.com/sqlitebrowser/sqlitebrowser/tree/openssl11_v1

Need to get some sleep atm, but a Win64 compile using the above branch is running now.  Hopefully it completes ok. :smile:

**Comment by karim** [comment]
> Getting the MSI installers to detect and correctly remove installers done using the previous .exe releases

I'm really sorry about that, I keep forgetting. :smiley:

I think the new version of SQLCipher, and OpenSSL, should have their own issues now.

**Comment by justinclift** [comment]
Good point.  Just created an issue for that (#1640).

Do you want to rename the `win*_new.bat` files to replace the older non `_new` ones, and close this issue? :smile:

**Comment by karim** [comment]
Sure. I was waiting for them to run for a couple of days to check if they are not breaking anything.

**Comment by justinclift** [comment]
Excellent.  Good stuff @karim. :smile:
